packages/headless/src/controllers/commerce/did-you-mean/headless-did-you-mean.ts

```diff
@@ -0,0 +1,69 @@
+import {
+  QueryCorrection,
+  WordCorrection,
+} from '../../../api/search/search/query-corrections';
+import {CommerceEngine} from '../../../app/commerce-engine/commerce-engine';
+import {CoreEngine} from '../../../app/engine';
+import {didYouMeanReducer as didYouMean} from '../../../features/commerce/did-you-mean/did-you-mean-slice';
+import {CommerceDidYouMeanSection} from '../../../state/state-sections';
+import {loadReducerError} from '../../../utils/errors';
+import {
+  buildController,
+  Controller,
+} from '../../controller/headless-controller';
+import {DidYouMeanState} from '../../did-you-mean/headless-did-you-mean';
+
+export type {QueryCorrection, WordCorrection, DidYouMeanState};
+
+export interface DidYouMean extends Controller {
+  /**
+   * A scoped and simplified part of the headless state that is relevant to the `DidYouMean` controller.
+   */
+  state: DidYouMeanState;
+}
+
+/**
+ * The DidYouMean controller is responsible for handling query corrections.
+ * When a query returns no result but finds a possible query correction, the controller automatically triggers a new
+ * query with the suggested term.
+ *
+ * @param engine - The headless commerce engine.
+ */
+export function buildDidYouMean(engine: CommerceEngine): DidYouMean {
+  if (!loadDidYouMeanReducers(engine)) {
+    throw loadReducerError;
+  }
+
+  const controller = buildController(engine);
+
+  const getState = () => engine.state;
+
+  return {
+    ...controller,
+
+    get state() {
+      const state = getState();
+
+      const hasQueryCorrection =
+        state.didYouMean.queryCorrection.correctedQuery !== '' ||
```

Comment:
We should create `did-you-mean-selectors.ts`, right next to `did-you-mean-slice.ts`, which does those state access + state computation.

We'll want to spread out that pattern everywhere in Headless.
example: https://github.com/coveo/ui-kit/pull/3907/files#diff-b553f42cfb93b76e28fbad45cb4b97aec6022ec9f3cb9949c97e0e2b6a29abf2R72-R109